Can't get update button to populate the AMWallet table.

eoliveros opened the pull request.

eoliveros committed:
added address_search.py

eoliveros committed:
Added the changes to models.py, utils.py, and templates/amdevice_list.html

eoliveros committed:
changed back to when it was supposed to execture the updating of the address

djpnewton commented on zapd/address_search.py:
```diff
@@ -0,0 +1,29 @@
+import requests
```
This file is not used and should be removed

eoliveros replied:
removed.

djpnewton commented on zapd/models.py:
```diff
@@ -218,6 +218,33 @@ def with_multiple_devices(cls, session):
+    def update_wallet_address(cls, session, user):
+        # update txs
+        limit = 100
+        ISSUER_ADDR = '3PCj4WTJ9abwYakwL4NBxiq1z4DmViFc5X3'
```
you should get this address from the asset id

an asset id is also the transaction id of the transaction that created the asset so you can get the 'sender' of that transaction to know the 'ISSUER_ADDR' (eg https://nodes.wavesnodes.com/transactions/info/9R3iLi4qGLVWKc16Tg98gmRvgg1usGEYd7SgC1W5D6HB)

eoliveros replied:
the ISSUER_ADDR is just a variable name. It used to be ZAP_WALLET. and I'm using the variable for pulling the transaction:
http://nodes.wavesnodes.com/transactions/address/<ISSUER_ADDR>/limit/<limit>
ie.
```
http://nodes.wavesnodes.com/transactions/address/3PCj4WTJ9abwYakwL4NBxiq1z4DmViFc5X3/limit/10
```
statically assigned the ISSUER_ADDR = ZAP_WALLET. Or do you want me to pull the records for the info and then grab the ISSUE_ADDR from it?

djpnewton replied:
yes i want you to get the issuer address dynamically so that it automatically gets the right issuer address based on the asset id

eoliveros replied:
should be in the commit as well.

djpnewton commented on zapd/templates/amdevice_list.html:
```diff
@@ -5,3 +5,8 @@
     <a href="multiple_devices" title="Show With Multiple Devices">Show With Multiple Devices</a>
 </li>
 {% endblock %}
+{% block model_menu_bar_after_filters %}
```
put this in the same block as 'multiple_devices'

eoliveros replied:
changed.

djpnewton commented on zapd/models.py:
```diff
@@ -218,6 +218,33 @@ def with_multiple_devices(cls, session):
         return session.query(cls).filter(cls.id.in_(ids))
 
+    @classmethod
+    def update_wallet_address(cls, session, user):
```
lets rename this to something like 'initial wallet addresses'

eoliveros replied:
changed to initial_wallet_addresses

djpnewton commented on zapd/models.py:
```diff
@@ -218,6 +218,33 @@ def with_multiple_devices(cls, session):
+                    break
+        for key, value in addrs:
+            address = key
+            session.add(AMWallet(address))
```
we also want to add a 'dummy' device associated with the address (just fill the values with 'n/a' or something), this means we will be able to detect when multiple devices create the same address

eoliveros replied:
so we need to create a new column in AMWallet called "device" and fill it with 'n/a' or 'unknown' or thinking of something else?

djpnewton replied:
no there is already a table called AMDevice, take a look here where we add an AMWallet object and an AMDevice object which is related to the wallet (https://github.com/djpnewton/zapd/blob/78ffaad3b60f15e5e6e40dc603f18685c8f115d5/zapd/web.py#L312)

eoliveros replied:
So I need to add a new record in the AMDevice table with "n/a" or "unknown" values for the following app_version, os, os_version, manufacturer, brand and device_id?

djpnewton replied:
yes

eoliveros replied:
ok, created the class method:
```
    @classmethod
    def from_device_id(cls, session, device_id):
        return session.query(cls).filter(cls.device_id == device_id).first()
```

then inside inside the app.py:
```
def create_device(device_id):
    devices = AMDevice.from_device_id(db.session, device_id)
    if not devices:
        devices = AMDevice(device_id)
    db.session.add(devices)
    return devices
```

keep getting the following:
```
Traceback (most recent call last):
  File "app.py", line 136, in <module>
    create_device("unknown")
  File "app.py", line 86, in create_device
    devices = AMDevice.from_device_id(db.session)
TypeError: from_device_id() missing 1 required positional argument: 'device_id'
```

missing something cant see it though.
which is then run iwith:
```
if __name__ == "__main__":
    setup_logging(logging.DEBUG)

    # create tables
    db.create_all()
    create_role("admin", "super user")
    create_role("proposer", "Can propose zap payments")
    create_role("authorizer", "Can authorize zap payments")
    create_category("marketing", "")
    create_category("misc", "")
    create_category("customer incentive", "")
    create_category("staff incentive", "")
    create_category("merchant incentive", "")
    create_category("testing", "")
    create_category("rebate", "")
    create_device("unknown")
```

eoliveros replied:
ok, edited the models.py
```
                    break
        for key, value in addrs:
            address = key
            wallet = address
            session.add(AMWallet(address))
            session.add(AMDevice(wallet, 'unknown app version', 'unknown os', 'unknown os version', 'unknown manufacturer', 'unknown brand', 'unknown device id')
```

Is that what you meant?

djpnewton replied:
no the address string is *not* a wallet object you need to do something like this:
```
for key, value in addrs:
    wallet = AMWallet(key)
    session.add(wallet)
    session.add(AMDevice(wallet, 'n/a', 'n/a', 'n/a', 'n/a', 'n/a', 'n/a')
```

eoliveros replied:
changed this to match.

djpnewton commented on zapd/models.py:
```diff
@@ -888,3 +915,10 @@ def is_accessible(self):
+    def update(self):
+        #print('Executing the update')
+        return_url = self.get_url('.index_view')
+        AMWallet.update_wallet_address(db.session)
```
if the table already has any entries then return an error and do not initialize the table data

use flask 'flash' functionality to add a success or error message

djpnewton reviewed (changes requested):
I have left a bunch of review comments for you to address

eoliveros committed:
deleted file

eoliveros committed:
Changed the template

eoliveros committed:
Changed the update_wallet_address to initial_wallet_addresses

eoliveros committed:
added new func for looking for the issuer_addr. changed classmethod to include the issuer_addr search

eoliveros committed:
fixed issue with issuer_addr in models.py

eoliveros committed:
removed @classmethod from_device_id in models.py and changed teh for loop to add devices for unknown.

djpnewton commented on zapd/models.py:
```diff
@@ -207,10 +211,6 @@ class AMWallet(db.Model):
         self.address = address
 
-    @classmethod
-    def from_address(cls, session, address):
```
you should not remove this code is it used in other parts

eoliveros replied:
accidentally removed it, i was supposed to remove from_device_id. placed from_address back,

djpnewton commented on zapd/models.py:
```diff
@@ -195,6 +195,10 @@ def __init__(self, wallet, app_version, os, os_version, manufacturer, brand, dev
         self.device_id = device_id
 
+    @classmethod
+    def from_device_id(cls, session, device_id):
```
this is not required

eoliveros replied:
This should be gone.

eoliveros committed:
fixed the mistake on model.py.

djpnewton commented on zapd/templates/amdevice_list.html:
```diff
@@ -4,4 +4,7 @@
     <a href="multiple_devices" title="Show With Multiple Devices">Show With Multiple Devices</a>
 </li>
+<li>
+    <a href="" title="Update Address List">Update</a>
```
href is not filled

eoliveros replied:
trying to figure out how I could get the "Update" button to run:
```
    @classmethod
    def initial_wallet_addresses(cls, session):
        # update txs
        limit = 100
        issuer_addr = issuer_address(app.config["NODE_ADDRESS"], app.config["ASSET_ID"])
        oldest_txid = None
        txs = []
        addrs = {}
        while True:
            have_tx = False
            txs = blockchain_transactions(app.config["NODE_ADDRESS"], issuer_addr, limit, oldest_txid)
            for tx in txs:
                oldest_txid = tx["id"]
                if tx["type"] == 4 and tx["assetId"] == app.config["ASSET_ID"]:
                    if 'sender' in tx:
                        sender = tx['sender']
                        addrs[sender] = 1
                    if 'recipient' in tx:
                        recipient = tx['recipient']
                        addrs[recipient] = 1
                if have_tx or len(txs) < limit:
                    break
        for key, value in addrs:
            wallet = AMWAllet(key)
            session.add(wallet)
            session.add(AMDevice(wallet, 'n/a', 'n/a', 'n/a', 'n/a', 'n/a', 'n/a'))
        session.commit()
```
Which is under the AMWallet(db.Model).

At the moment, I have "@expose('update_address_list')" going to a webpage which has :
```
    <a href="" title="Update Address List">Update</a>
```

So its currently not doing anything. The question is how to get from the webpage to the AMWallet(db.Model), If I can get this to work, it should populate the table AMWallet address table and attach it to a 'n/a' device.

djpnewton replied:
href still is empty (https://www.w3schools.com/tags/att_a_href.asp)

eoliveros replied:
Placing:
```
<li>
    <a href="/update_list" title="Update Address List">Update</a>
</li>
```

Which means i need to put something in the web.py. The question then would be what do i put in the web.py:
```
@app.route("/update_list", methods=["POST"])
def update_list():
     this is what im trying to figure out. Need something in here that would use that section in the @classmodel for performing the query to loop through the blockchain and im not able to figure it out.
    return "ok"
```

djpnewton replied:
no we are using the flask admin expose (https://flask-admin.readthedocs.io/en/latest/api/mod_base/#flask_admin.base.expose) decorator which sets up the route for us

you have:
`@expose("/update_address_list")`

so why dont you put:
`<a href="/update_address_list" ...`

eoliveros replied:
ive changed the naming i used local repo to match remote repo.

Still ahve the problem of what i need to put in the web.py:
```
@app.route("/update_list", methods=["POST"])
def update_list():
     ###this is what im trying to figure out. Need something in here that would use that section in the @classmodel for performing the query to loop through the blockchain and im not able to figure it out.
    return "ok"
```

Currently its returning "ok" on my screen. 

changing it to make it look like this:
```
@app.route("/update_address_list", methods=["GET"])
def update_address_list():
    AMWallet.initial_wallet_addresses(db.session)

    return 'ok'
```

And now getting an error:
```
[2020-06-12 01:17:13,288] ERROR in app: Exception on /update_address_list [GET]
Traceback (most recent call last):
  File "/root/git/zapd/zapd/venv/lib/python3.6/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/root/git/zapd/zapd/venv/lib/python3.6/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/root/git/zapd/zapd/venv/lib/python3.6/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/root/git/zapd/zapd/venv/lib/python3.6/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/root/git/zapd/zapd/venv/lib/python3.6/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/root/git/zapd/zapd/venv/lib/python3.6/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/root/git/zapd/zapd/web.py", line 325, in update_address_list
    AMWallet.initial_wallet_addresses(db.session)
  File "/root/git/zapd/zapd/models.py", line 225, in initial_wallet_addresses
    issuer_addr = issuer_address(app.config["NODE_ADDRESS"], app.config["ASSET_ID"])
KeyError: 'NODE_ADDRESS'
```

So it looks like i can't use app.config["NODE_ADDRESS"] from the models.py

This is the @classmethod fromo the models.py:
```
    @classmethod
    def initial_wallet_addresses(cls, session):
        # update txs
        limit = 100
        issuer_addr = issuer_address(app.config["NODE_ADDRESS"], app.config["ASSET_ID"])
        oldest_txid = None
        txs = []
        addrs = {}
        while True:
            have_tx = False
            txs = blockchain_transactions(app.config["NODE_ADDRESS"], issuer_addr, limit, oldest_txid)
            for tx in txs:
                oldest_txid = tx["id"]
                if tx["type"] == 4 and tx["assetId"] == app.config["ASSET_ID"]:
                    if 'sender' in tx:
                        sender = tx['sender']
                        addrs[sender] = 1
                    if 'recipient' in tx:
                        recipient = tx['recipient']
                        addrs[recipient] = 1
                if have_tx or len(txs) < limit:
                    break
        for key, value in addrs:
            wallet = AMWAllet(key)
            session.add(wallet)
            session.add(AMDevice(wallet, 'n/a', 'n/a', 'n/a', 'n/a', 'n/a', 'n/a'))
        session.commit()
```
Which most likely going to fail for the other app.config["ASSET_ID"] as well.

Any ideas on how I could use app.config in the models.py?

eoliveros replied:
since app.config["NODE_ADDRESS"] and app.config["ASSET_ID"] is not in the script, im static placing it in the models.py

checking to see what else we need/testing.

eoliveros replied:
managed to get the following in the console:
```
:: requesting http://nodes.wavesnodes.com/transactions/address/3PCj4WTJ9abwYakwL4NBxiq1z4DmViFc5X3/limit/1000?after=3G6TvYnwmfQj2fhSKggUwHYjHGaQTLgSbX6p2ueQHnpd..
:: retrieved 1000 records
1000
:: requesting http://nodes.wavesnodes.com/transactions/address/3PCj4WTJ9abwYakwL4NBxiq1z4DmViFc5X3/limit/1000?after=J9RGNURR7PkghXdkVWu3oekfEfxcd87uFvtDLYrRFRfj..
:: retrieved 1000 records
1000
:: requesting http://nodes.wavesnodes.com/transactions/address/3PCj4WTJ9abwYakwL4NBxiq1z4DmViFc5X3/limit/1000?after=9BMiuGTDiCjzeXSV7fh8MRGaKugpeSN1zU6LPaTbU3AH..
:: retrieved 1000 records
1000
:: requesting http://nodes.wavesnodes.com/transactions/address/3PCj4WTJ9abwYakwL4NBxiq1z4DmViFc5X3/limit/1000?after=Hd2CDyAEJBcFKD4gqDEGcFDDTsXr6jnQx8AHcnC8cy8D..
:: retrieved 1000 records
1000
:: requesting http://nodes.wavesnodes.com/transactions/address/3PCj4WTJ9abwYakwL4NBxiq1z4DmViFc5X3/limit/1000?after=46DQCkg61PwhWsAkZmVM5tMSRYBeJgakKYCrfRSHw9Zq..
:: retrieved 1000 records
1000
:: requesting http://nodes.wavesnodes.com/transactions/address/3PCj4WTJ9abwYakwL4NBxiq1z4DmViFc5X3/limit/1000?after=8twKCvFHeK6L1heBaeSSQGFUrQcV6S5xTfV9iLdJLA34..
:: retrieved 836 records
1
:: requesting http://nodes.wavesnodes.com/transactions/address/3PCj4WTJ9abwYakwL4NBxiq1z4DmViFc5X3/limit/1000?after=8BnLfbbh8K5qsVuPus2DTCe3RNvJYiBJXwCxUj4t85u..
:: retrieved 835 records
1
```

Not sure if this a ratelimit of some sort? Or maybe that if the no. of records is less than the limit its starts requesting the exact same query with the next id in the last result.

Need to look at it further.

eoliveros replied:
so my one appears to run the blockchain_transaction and instead of looping in there, it would go out of the function and then go back in.

the question is why is it remaining in the loop and not breaking when the no. of transaction is below the limit.

need to find out why this is and force it break afterwards so i could look at writing the result to the db.

eoliveros replied:
so the loop is not breaking because I'm trying to break from inside the if statement. the moment I placed that in the same level as the for loop it breaks. Which is what I wanted. Checking to see if it would now write to the sqlite db. dropping the tables and then re-run the app.

eoliveros replied:
pulled the changes and it looks like its working:
![image](https://user-images.githubusercontent.com/47799652/84608156-2d6c7000-af05-11ea-940d-447abbf6de8c.png)

running the initialize again to see the result:
![image](https://user-images.githubusercontent.com/47799652/84608188-51c84c80-af05-11ea-921a-5667c39560ad.png)


Looks like its working. Pushing this to my remote.

eoliveros committed:
cahnged.

djpnewton committed:
cleaned up wallet addresses initialization
 - added error handling
 - only can initialize addressess on first install
 - fixed endpoint links/names

djpnewton commented on zapd/utils.py:
```diff
@@ -122,6 +122,64 @@ def is_address(s):
     except:
         return False
 
+def initial_wallet_addresses(cls, session):
```
this is a duplicated unused function

eoliveros replied:
removed

djpnewton commented on zapd/utils.py:
```diff
@@ -122,6 +122,64 @@ def is_address(s):
+    
+    return txs_result
+
+def testing1():
```
unused code

eoliveros replied:
Removed

djpnewton reviewed (changes requested):
just need to removed the unused code and then I will merge

eoliveros committed:
Removed unused code
